dlrover/trainer/torch/elastic/checkpoint.py

```diff
@@ -0,0 +1,219 @@
+
+import torch
+import torch.distributed as dist
+from torch.distributed.fsdp import FullStateDictConfig
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp import StateDictType
+
+from dlrover.python.common.log import default_logger as logger
+from dlrover.trainer.torch.elastic.sampler import ElasticDistributedSampler
+
+CKPT_DIR_PREFIX = "checkpoint-"
+
+
+def init_dir(dir):
+    if os.path.exists(dir):
+        shutil.rmtree(dir)
+    os.makedirs(dir)
+
+
+def sync():
+    if dist.is_initialized():
+        dist.barrier()
+
+
+class CheckpointManger(object):
+    """CheckpontManager can save and load checkpoint states.
+
+    Args:
+        model (nn.Module): an instance of `torch.nn.Module`.
+        optimizer (Optimizer): an instance of `torch.optim.Optimizer`.
+        dataloader (DataLader): an instance of `torch.utils.data.DataLoader`.
+            The sampler of DataLoader should be an instance of
+            `dlrover.trainer.torch.elastic.ElasticDistribuedSampler`.
+        directory (str): the directory to save the checkpoint states.
+        rank (int): the rank of process in the communication world.
+        max_to_keep (int): the max number of checkpoint to keep. The oldest
+            checkpoint files will be removed if the number of checkpoints
+            is bigger than max_to_kepp.
+
+    Example::
+        >>> ckpt_manager = CheckpointManager(
+        >>>    model, optimizer, train_dataloader, "/tmp/checkpoint/"
+        >>> )
+        >>> ckpt_manager.save(0, 10)
+        >>> ckpt_manger.load()
+
+    """
+
+    def __init__(
+        self,
+        model,
+        optimizer,
+        dataloader,
+        directory,
+        rank=0,
+        max_to_keep=None,
+    ):
+        self.model = model
+        self.optimizer = optimizer
+        self.dataloader = dataloader
+        self.directory = directory
+        self.rank = rank
+        self.max_to_keep = max_to_keep
+
+    def log_rank0(self, log):
+        if self.rank == 0:
+            logger.info(log)
+
+    def _is_rank0(self):
+        return self.rank == 0
+
+    def save(self, epoch, step):
+        """
+        Save the checkpoint of model, optimizer, dataloader into the directory
+        `{self.directory}/checkpoint-{step}/`. If use_fdsp, all ranks will save
+        the part of the model and optimizer states into the file
+        `checkpoint-{step}/part-{rank}.pt`. If not use_fsdp, only rank0 saves
+        the total model and optimizer states into the file
+        `checkpoint-{step}/checkpoint.pt`.
+
+        Args:
+            epoch (int): the epoch index.
+            step (int): the iteration step in the epoch.
+        """
+        self.log_rank0(f"Save checkpoint of step={step} of epoch={epoch}.")
+        step = step + epoch * len(self.dataloader)
+        is_fsdp_model = isinstance(self.model, FSDP)
+        msd, osd = get_model_optim_state(
+            self.model, self.optimizer, is_fsdp_model
```

Comment:
What if the model is wrapped by `DeepSpeedEngine`?

Reply:
I have refactored the implementation and developed independent classes for local, DDP and FSDP. If the model is wrapped by `DeepSpeedEngine`, we will raise a `NotImplementedError`. Later, we can implement the `CheckpointManager` for `DeepSpeedEngine`.